### PR TITLE
Table: USN

### LINKED ISSUE #66: Table: USN
## Description

A table tracking USN metrics on NEAR

## TODO
- [x]  Design the schema for this table
- [x]  Get confirmation on schema
- [x]  Build schema
- [x]  Schema build review
- [x]  Submit Pull Request

### PR DESCRIPTION
**Pull request by geewynn**
# Description
Created both silver and core ez data model and table for USN transactions
closes #66 

# Tests

- [x] Please provide evidence of your successful `dbt run` / `dbt test` here
- [x] Any comparison between `prod` and `dev` for any schema change
**Silver usn**
```
**dbt run**
root@72df438d7633:/near# dbt test --select silver__usn
23:37:14  Running with dbt=1.0.0
23:37:14  Found 33 models, 511 tests, 0 snapshots, 0 analyses, 628 macros, 1 operation, 0 seed files, 8 sources, 0 exposures, 0 metrics
23:37:14  
23:37:24  
23:37:24  Running 1 on-run-start hook
23:37:24  1 of 1 START hook: near.on-run-start.0.......................................... [RUN]
23:37:24  1 of 1 OK hook: near.on-run-start.0............................................. [OK in 0.00s]
23:37:24  
23:37:24  Concurrency: 4 threads (target='dev')
23:37:24  
23:37:24  1 of 17 START test dbt_expectations_expect_column_values_to_be_in_type_list_silver__usn_AMOUNT__NUMBER__FLOAT [RUN]
23:37:24  2 of 17 START test dbt_expectations_expect_column_values_to_be_in_type_list_silver__usn_BLOCK_ID__NUMBER__FLOAT [RUN]
23:37:24  3 of 17 START test dbt_expectations_expect_column_values_to_be_in_type_list_silver__usn_BLOCK_TIMESTAMP__TIMESTAMP_NTZ [RUN]
23:37:24  4 of 17 START test dbt_expectations_expect_column_values_to_be_in_type_list_silver__usn_METHOD_NAMES__STRING__VARCHAR [RUN]
23:37:29  4 of 17 PASS dbt_expectations_expect_column_values_to_be_in_type_list_silver__usn_METHOD_NAMES__STRING__VARCHAR [PASS in 5.46s]
23:37:29  3 of 17 PASS dbt_expectations_expect_column_values_to_be_in_type_list_silver__usn_BLOCK_TIMESTAMP__TIMESTAMP_NTZ [PASS in 5.47s]
23:37:29  5 of 17 START test dbt_expectations_expect_column_values_to_be_in_type_list_silver__usn_NEW_OWNER__STRING__VARCHAR [RUN]
23:37:29  2 of 17 PASS dbt_expectations_expect_column_values_to_be_in_type_list_silver__usn_BLOCK_ID__NUMBER__FLOAT [PASS in 5.48s]
23:37:29  6 of 17 START test dbt_expectations_expect_column_values_to_be_in_type_list_silver__usn_OLD_OWNER__STRING__VARCHAR [RUN]
23:37:29  7 of 17 START test dbt_expectations_expect_column_values_to_be_in_type_list_silver__usn_TX_HASH__STRING__VARCHAR [RUN]
23:37:29  1 of 17 PASS dbt_expectations_expect_column_values_to_be_in_type_list_silver__usn_AMOUNT__NUMBER__FLOAT [PASS in 5.53s]
23:37:29  8 of 17 START test dbt_expectations_expect_column_values_to_be_in_type_list_silver__usn_TX_RECEIVER__STRING__VARCHAR [RUN]
23:37:34  6 of 17 PASS dbt_expectations_expect_column_values_to_be_in_type_list_silver__usn_OLD_OWNER__STRING__VARCHAR [PASS in 4.62s]
23:37:34  9 of 17 START test dbt_expectations_expect_column_values_to_be_in_type_list_silver__usn_TX_SIGNER__STRING__VARCHAR [RUN]
23:37:34  7 of 17 PASS dbt_expectations_expect_column_values_to_be_in_type_list_silver__usn_TX_HASH__STRING__VARCHAR [PASS in 4.63s]
23:37:34  10 of 17 START test dbt_utils_unique_combination_of_columns_silver__usn_tx_hash. [RUN]
23:37:34  5 of 17 PASS dbt_expectations_expect_column_values_to_be_in_type_list_silver__usn_NEW_OWNER__STRING__VARCHAR [PASS in 4.76s]
23:37:34  11 of 17 START test not_null_silver__usn_AMOUNT................................. [RUN]
23:37:34  8 of 17 PASS dbt_expectations_expect_column_values_to_be_in_type_list_silver__usn_TX_RECEIVER__STRING__VARCHAR [PASS in 4.95s]
23:37:34  12 of 17 START test not_null_silver__usn_BLOCK_ID............................... [RUN]
23:37:38  12 of 17 PASS not_null_silver__usn_BLOCK_ID..................................... [PASS in 4.12s]
23:37:38  13 of 17 START test not_null_silver__usn_BLOCK_TIMESTAMP........................ [RUN]
23:37:38  11 of 17 PASS not_null_silver__usn_AMOUNT....................................... [PASS in 4.60s]
23:37:38  10 of 17 PASS dbt_utils_unique_combination_of_columns_silver__usn_tx_hash....... [PASS in 4.71s]
23:37:38  14 of 17 START test not_null_silver__usn_METHOD_NAMES........................... [RUN]
23:37:38  15 of 17 START test not_null_silver__usn_TX_HASH................................ [RUN]
23:37:39  9 of 17 PASS dbt_expectations_expect_column_values_to_be_in_type_list_silver__usn_TX_SIGNER__STRING__VARCHAR [PASS in 5.13s]
23:37:39  16 of 17 START test not_null_silver__usn_TX_RECEIVER............................ [RUN]
23:37:41  13 of 17 PASS not_null_silver__usn_BLOCK_TIMESTAMP.............................. [PASS in 3.29s]
23:37:42  17 of 17 START test not_null_silver__usn_TX_SIGNER.............................. [RUN]
23:37:43  14 of 17 PASS not_null_silver__usn_METHOD_NAMES................................. [PASS in 4.19s]
23:37:43  15 of 17 PASS not_null_silver__usn_TX_HASH...................................... [PASS in 4.24s]
23:37:45  16 of 17 PASS not_null_silver__usn_TX_RECEIVER.................................. [PASS in 5.96s]
23:37:48  17 of 17 PASS not_null_silver__usn_TX_SIGNER.................................... [PASS in 6.31s]
23:37:48  
23:37:48  Finished running 17 tests, 1 hook in 33.79s.
23:37:48  
23:37:48  Completed successfully
23:37:48  
23:37:48  Done. PASS=17 WARN=0 ERROR=0 SKIP=0 TOTAL=17
```
```
**dbt test**
root@72df438d7633:/near# dbt run --select silver__usn
23:51:08  Running with dbt=1.0.0
23:51:09  Found 34 models, 528 tests, 0 snapshots, 0 analyses, 628 macros, 1 operation, 0 seed files, 8 sources, 0 exposures, 0 metrics
23:51:09  
23:51:18  
23:51:18  Running 1 on-run-start hook
23:51:18  1 of 1 START hook: near.on-run-start.0.......................................... [RUN]
23:51:18  1 of 1 OK hook: near.on-run-start.0............................................. [OK in 0.00s]
23:51:18  
23:51:18  Concurrency: 4 threads (target='dev')
23:51:18  
23:51:18  1 of 1 START incremental model silver.usn....................................... [RUN]
23:51:28  1 of 1 OK created incremental model silver.usn.................................. [SUCCESS 1 in 9.69s]
23:51:28  
23:51:28  Finished running 1 incremental model, 1 hook in 19.19s.
23:51:28  
23:51:28  Completed successfully
23:51:28  
23:51:28  Done. PASS=1 WARN=0 ERROR=0 SKIP=0 TOTAL=1
```

**Core EZ USN**
```
**dbt run**
root@72df438d7633:/near# dbt run --select core__ez_usn
23:40:17  Running with dbt=1.0.0
23:40:18  Found 34 models, 528 tests, 0 snapshots, 0 analyses, 628 macros, 1 operation, 0 seed files, 8 sources, 0 exposures, 0 metrics
23:40:18  
23:40:29  
23:40:29  Running 1 on-run-start hook
23:40:29  1 of 1 START hook: near.on-run-start.0.......................................... [RUN]
23:40:29  1 of 1 OK hook: near.on-run-start.0............................................. [OK in 0.00s]
23:40:29  
23:40:29  Concurrency: 4 threads (target='dev')
23:40:29  
23:40:29  1 of 1 START view model core.ez_usn............................................. [RUN]
23:40:32  1 of 1 OK created view model core.ez_usn........................................ [SUCCESS 1 in 3.03s]
23:40:32  
23:40:32  Finished running 1 view model, 1 hook in 14.03s.
23:40:32  
23:40:32  Completed successfully
23:40:32  
23:40:32  Done. PASS=1 WARN=0 ERROR=0 SKIP=0 TOTAL=1
```
```
**dbt test**
root@72df438d7633:/near# dbt test --select core__ez_usn
23:40:47  Running with dbt=1.0.0
23:40:47  Found 34 models, 528 tests, 0 snapshots, 0 analyses, 628 macros, 1 operation, 0 seed files, 8 sources, 0 exposures, 0 metrics
23:40:47  
23:40:54  
23:40:54  Running 1 on-run-start hook
23:40:54  1 of 1 START hook: near.on-run-start.0.......................................... [RUN]
23:40:54  1 of 1 OK hook: near.on-run-start.0............................................. [OK in 0.00s]
23:40:54  
23:40:54  Concurrency: 4 threads (target='dev')
23:40:54  
23:40:54  1 of 17 START test dbt_expectations_expect_column_values_to_be_in_type_list_core__ez_usn_AMOUNT__NUMBER__FLOAT [RUN]
23:40:54  2 of 17 START test dbt_expectations_expect_column_values_to_be_in_type_list_core__ez_usn_BLOCK_ID__NUMBER__FLOAT [RUN]
23:40:54  3 of 17 START test dbt_expectations_expect_column_values_to_be_in_type_list_core__ez_usn_BLOCK_TIMESTAMP__TIMESTAMP_NTZ [RUN]
23:40:54  4 of 17 START test dbt_expectations_expect_column_values_to_be_in_type_list_core__ez_usn_METHOD_NAMES__STRING__VARCHAR [RUN]
23:41:00  4 of 17 PASS dbt_expectations_expect_column_values_to_be_in_type_list_core__ez_usn_METHOD_NAMES__STRING__VARCHAR [PASS in 5.27s]
23:41:00  5 of 17 START test dbt_expectations_expect_column_values_to_be_in_type_list_core__ez_usn_NEW_OWNER__STRING__VARCHAR [RUN]
23:41:00  1 of 17 PASS dbt_expectations_expect_column_values_to_be_in_type_list_core__ez_usn_AMOUNT__NUMBER__FLOAT [PASS in 5.77s]
23:41:00  6 of 17 START test dbt_expectations_expect_column_values_to_be_in_type_list_core__ez_usn_OLD_OWNER__STRING__VARCHAR [RUN]
23:41:01  2 of 17 PASS dbt_expectations_expect_column_values_to_be_in_type_list_core__ez_usn_BLOCK_ID__NUMBER__FLOAT [PASS in 6.17s]
23:41:01  7 of 17 START test dbt_expectations_expect_column_values_to_be_in_type_list_core__ez_usn_TX_HASH__STRING__VARCHAR [RUN]
23:41:01  3 of 17 PASS dbt_expectations_expect_column_values_to_be_in_type_list_core__ez_usn_BLOCK_TIMESTAMP__TIMESTAMP_NTZ [PASS in 6.56s]
23:41:01  8 of 17 START test dbt_expectations_expect_column_values_to_be_in_type_list_core__ez_usn_TX_RECEIVER__STRING__VARCHAR [RUN]
23:41:03  5 of 17 PASS dbt_expectations_expect_column_values_to_be_in_type_list_core__ez_usn_NEW_OWNER__STRING__VARCHAR [PASS in 3.55s]
23:41:03  9 of 17 START test dbt_expectations_expect_column_values_to_be_in_type_list_core__ez_usn_TX_SIGNER__STRING__VARCHAR [RUN]
23:41:04  6 of 17 PASS dbt_expectations_expect_column_values_to_be_in_type_list_core__ez_usn_OLD_OWNER__STRING__VARCHAR [PASS in 3.50s]
23:41:04  10 of 17 START test dbt_utils_unique_combination_of_columns_core__ez_usn_tx_hash [RUN]
23:41:04  7 of 17 PASS dbt_expectations_expect_column_values_to_be_in_type_list_core__ez_usn_TX_HASH__STRING__VARCHAR [PASS in 3.15s]
23:41:04  11 of 17 START test not_null_core__ez_usn_AMOUNT................................ [RUN]
23:41:04  8 of 17 PASS dbt_expectations_expect_column_values_to_be_in_type_list_core__ez_usn_TX_RECEIVER__STRING__VARCHAR [PASS in 3.38s]
23:41:04  12 of 17 START test not_null_core__ez_usn_BLOCK_ID.............................. [RUN]
23:41:07  10 of 17 PASS dbt_utils_unique_combination_of_columns_core__ez_usn_tx_hash...... [PASS in 3.52s]
23:41:07  11 of 17 PASS not_null_core__ez_usn_AMOUNT...................................... [PASS in 3.49s]
23:41:07  13 of 17 START test not_null_core__ez_usn_BLOCK_TIMESTAMP....................... [RUN]
23:41:07  14 of 17 START test not_null_core__ez_usn_METHOD_NAMES.......................... [RUN]
23:41:07  9 of 17 PASS dbt_expectations_expect_column_values_to_be_in_type_list_core__ez_usn_TX_SIGNER__STRING__VARCHAR [PASS in 3.99s]
23:41:07  15 of 17 START test not_null_core__ez_usn_TX_HASH............................... [RUN]
23:41:08  12 of 17 PASS not_null_core__ez_usn_BLOCK_ID.................................... [PASS in 3.09s]
23:41:08  16 of 17 START test not_null_core__ez_usn_TX_RECEIVER........................... [RUN]
23:41:12  15 of 17 PASS not_null_core__ez_usn_TX_HASH..................................... [PASS in 4.33s]
23:41:12  16 of 17 PASS not_null_core__ez_usn_TX_RECEIVER................................. [PASS in 4.06s]
23:41:12  17 of 17 START test not_null_core__ez_usn_TX_SIGNER............................. [RUN]
23:41:12  14 of 17 PASS not_null_core__ez_usn_METHOD_NAMES................................ [PASS in 4.44s]
23:41:12  13 of 17 PASS not_null_core__ez_usn_BLOCK_TIMESTAMP............................. [PASS in 4.52s]
23:41:15  17 of 17 PASS not_null_core__ez_usn_TX_SIGNER................................... [PASS in 3.49s]
23:41:15  
23:41:15  Finished running 17 tests, 1 hook in 28.11s.
23:41:15  
23:41:15  Completed successfully
23:41:15  
23:41:15  Done. PASS=17 WARN=0 ERROR=0 SKIP=0 TOTAL=17
```

# Checklist
- [x] Follow [dbt style guide](https://github.com/dbt-labs/corp/blob/main/dbt_style_guide.md)
- [x] Tag the person(s) responsible for reviewing proposed changes
- [x] Notes to deployment, if a `full-refresh` is needed for any table
- [x] Run `git merge main` to pull any changes from remote into your branch prior to merge.

### CODE REVIEW

**Comment by 0xissui** [comment]
Good stuff! A thought, perhaps we should have a decimally adjusted for the EZ view

File: models/core/core__ez_usn.sql [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+        {{ ref('silver__usn') }}
+)
+SELECT
+    block_timestamp,
+    block_id,
+    method_names,
+    tx_hash,
+    tx_receiver,
+    tx_signer,
+    amount
```
Comment (forgxyz) [inline models/core/core__ez_usn.sql:20]: For `ez_` views, let's decimal adjust any amount so users can take the number presented as actual.

File: models/silver/silver__usn.sql [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        TRY_PARSE_JSON(
+            TRY_BASE64_DECODE_STRING(
+                tx: actions[0]: FunctionCall: args
+            )
+        ) AS parsed_arg,
+        tx: actions[0]: FunctionCall: method_name :: STRING AS method_names,
+        tx,
+        _ingested_at,
+        _inserted_timestamp
+    FROM
```
Comment (forgxyz) [inline models/silver/silver__usn.sql:33]: Include tx status, if it succeeded or failed, in the CTEs (only commenting in one place, here)

File: models/silver/silver__usn.sql [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+),
+buy AS (
+    SELECT
+        block_timestamp,
+        block_id,
+        method_names,
+        tx_hash,
+        tx_receiver,
+        tx_signer,
+        tx: actions[0]: FunctionCall: deposit :: NUMBER AS amount,
```
Comment (forgxyz) [inline models/silver/silver__usn.sql:47]: Do we know if the amount here is listed in USN or with the amount of NEAR or USDT used to purchase/mint?

File: models/silver/silver__usn.sql [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+),
+sell AS (
+    SELECT
+        block_timestamp,
+        block_id,
+        method_names,
+        tx_hash,
+        tx_receiver,
+        tx_signer,
+        parsed_arg: amount :: NUMBER AS amount,
```
Comment (forgxyz) [inline models/silver/silver__usn.sql:63]: Same as above, can we confirm this amount is always USN, and not NEAR or another token.

File: models/core/core__ez_usn.sql (hunk 1 above)
Comment (forgxyz) [reply models/core/core__ez_usn.sql:20]: In tests transactions I just made, the buy was 10^6 but the sell was 10^18... I would confirm that it's always that way.
It was USDT<-> NEAR, so I am assuming the decimal reflects the asset.
If the alt asset was previously NEAR, the decimal adjustment may be different.

File: models/descriptions/amount.md [hunk 5]
```diff
@@ -0,0 +1,5 @@
+{% docs amount %}
+
+The amount of tokens transferred
```
Comment (forgxyz) [inline models/descriptions/amount.md:3]: Add something along the lines of:
"If in the `ez_` view, this number is decimal adjusted to actual amount. If in `silver` level or a `fact_` view, the raw amount is presented.

**Review by forgxyz — CHANGES_REQUESTED** [review]

**Comment by forgxyz** [comment]
Finally, I would recommend renaming the models to be a bit more descriptive of the data they reflect. Possibly `usn_supply` ?

**Comment by forgxyz** [comment]
> Good stuff! A thought, perhaps we should have a decimally adjusted for the EZ view

Agree! Didn't see your comment before my review.
`ez_` tables should be fully curated and as easy as possible to use (hence the naming).

File: models/core/core__ez_usn.sql (hunk 1 above)
Comment (geewynn) [reply models/core/core__ez_usn.sql]: Fixed this, now using usn amount gotten from the event logs. I confirmed that it was the right amount from the near block explorer.

File: models/silver/silver__usn.sql (hunk 2 above)
Comment (geewynn) [reply models/silver/silver__usn.sql]: Status isn't mentioned in the tx json.

File: models/silver/silver__usn.sql (hunk 3 above)
Comment (geewynn) [reply models/silver/silver__usn.sql]: It has now been modified to only use the usn values.

File: models/descriptions/amount.md (hunk 5 above)
Comment (geewynn) [reply models/descriptions/amount.md:3]: Done

**Review by forgxyz — APPROVED** [review]

**Pull request merged**